### PR TITLE
Support macro calls in type position

### PR DESCRIPTION
**Pull request by regiontog**
A [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fdc6dd4ddaece92a72fa2a292b75e27c) demonstrating the syntax in question.

### CODE REVIEW

File: crates/ra_syntax/src/grammar/types.rs [hunk 1]
```diff
@@ -243,6 +243,28 @@ pub(super) fn path_type(p: &mut Parser) {
     path_type_(p, true)
 }
 
+// test macro_call_type
+// type A = foo!();
+// type B = crate::foo!();
+fn path_or_macro_type_(p: &mut Parser, allow_bounds: bool) {
+    assert!(paths::is_path_start(p) || p.at(L_ANGLE));
+    let m = p.start();
+    paths::type_path(p);
```
Comment (regiontog) [inline crates/ra_syntax/src/grammar/types.rs]: This accepts paths with `Mode::Type` for path types and macro calls, but macro calls should really have `Mode::Use` AFAIK.

File: crates/ra_syntax/src/grammar/types.rs [hunk 2]
```diff
@@ -243,6 +243,28 @@ pub(super) fn path_type(p: &mut Parser) {
...
+
+    let kind = if p.at(EXCL) {
+        items::macro_call_after_excl(p);
+        MACRO_CALL
+    } else {
+        PATH_TYPE
+    };
+
+    if allow_bounds && p.eat(PLUS) {
```
[the remark this answers is not in the document]
Comment (DJMcNab) [reply crates/ra_syntax/src/grammar/types.rs]: I think this handles cases such as
```rust
struct X<T: foo!() +Debug>();
```

Im not quite sure why this is handled here and not in the bounds function though (can't easily look on mobile)

[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_syntax/src/grammar/types.rs]: Oh, indeed, I haven't hand my pu'er yet :D I guess this nonetheless means that inline test will be welcome here!

[the remark this answers is not in the document]
Comment (regiontog) [reply crates/ra_syntax/src/grammar/types.rs]: [Unfortunately rustc does not like macro calls in type parameter bounds](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4911531a1ceefc0d08483c9aacee01f6). Is there any other cases where allow_bounds is true? If not then maybe macros should not parse if allow_bounds is true. On the other hand it seems like reasonable syntax to me.

**Pull request merged**